Fix regex in resolve-value.js to allow nested CSS functions

juliovedovatto opened the pull request:
This fix will make work, when using nested CSS functions in the same declaration.

See https://github.com/MadLittleMods/postcss-css-variables/issues/96

MadLittleMods commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "postcss-css-variables",
-  "version": "0.12.0",
+  "version": "0.12.1",
```
Remove this version bump in the pull request. We will bump it when we push out a release via `npm version` command

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -10,7 +10,8 @@ var cloneSpliceParentOntoNodeWhen = require('./clone-splice-parent-onto-node-whe
 // See: http://dev.w3.org/csswg/css-variables/#funcdef-var
-var RE_VAR_FUNC = (/var\(\s*(--[^,\s]+?)(?:\s*,\s*(.+))?\s*\)/);
+//var RE_VAR_FUNC = (/var\(\s*(--[^,\s]+?)(?:\s*,\s*(.+))?\s*\)/);
+var RE_VAR_FUNC = (/var\(\s*(--[^,\s]+?)(?:\s*,\s*([^)]+))?\s*\)/);
```
It looks like this change breaks some existing tests (see CI result), https://travis-ci.org/MadLittleMods/postcss-css-variables/builds/520981414

We should also add some test cases for the thing you are trying to fix so we don't regress in the future

juliovedovatto commented:
First, thank you @MadLittleMods  for the opportunity to let me help with this small fix 😄

**My Scenario:**

Previously, I just did a hotfix, without testing it properly, because deadline was short. Due IE11 browser, we had to convert every `var()` to its own original value. We used https://material.io grid layout and it contains A LOT of CSS variables.

**My new merge proposal:**

I decided to change my approach, and improved the `RegExp` sentence. I added an extra instruction to help match nested functions with `var()` declarations. With a fallback using the original regex block.

In addition, I had to change your approach on how replace results. I added a `while` block (not sure if is a good idea), to replace until no `var()` is found. 

To finish, I just updated `nested-inside-other-func` test, with two situations: one with a nested function with a declared `var()` and another with a undefined `var()`. Using the orginal `RegExp` it fails, using my proposal, it passes 🎉

Please check if my logic makes sense to you and if you can imagine any extra scenario to test.

Cheers from Brazil 🇧🇷

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -41,66 +41,69 @@ var resolveValue = function(decl, map, /*optional*/ignorePseudoScope, /*internal
-				(!(matchingVarDeclMapItem || {}).isImportant || varDeclMapItem.isImportant)
-			) {
-				matchingVarDeclMapItem = varDeclMapItem;
+	while (resultantValue.match(new RegExp(RE_VAR_FUNC.source, 'g'))) {
```
Can you fill me in on why the `resultantValue.replace(new RegExp(RE_VAR_FUNC.source, 'g')` is no longer suitable?

juliovedovatto replied:
the intention was to create a recursive match, to make sure to replace all variables correctly.

the previous version was not working with some nested css functions.

MadLittleMods replied:
The recursive stuff should come from the `resolveValue` calls inside 🤔 (see below)

juliovedovatto replied:
In my previous approaches, the original `RegExp` used was not covering this case, for example:

Original

```css
.box-foo {
	width: calc(58.3333333333% - var(--some-width, 100px)); 
}
```

Result: 

```css
.box-foo {
width: calc(58.3333333333% - var(--some-width, 100px); 
}
```

---

I changed the previous regex to `var RE_VAR_FUNC = (/var\(\s*(--[^,\s]+?)(?:\s*,\s*([^()]+|.+))?\s*\)/);
`, it solved the case above, but it wasn't working now for this case:

```
.box-foo {
	width: calc(58.3333333333% - var(--missing, var(--some-width, 100px))); 
}
```
Since the regex now was treating each `var()` reparately, I had to force recursive match.

So... I though to use a `while` block. Not sure if was a clever idea.

---

But I decided to change the regex, to match fill only the `variablesUsedInValue` array and use `match-recursive` lib to help me to match  recursively.  Right now I'm not sure if the `while` block is relevant or not.

It makes sense to you?

MadLittleMods replied:
It seems like the recursive bits are duplicated now,

Here is where the `var()` in the fallback should be resolved, https://github.com/MadLittleMods/postcss-css-variables/blob/fe5c821b8feedbe80d8f4399a2b80ca166bb3515/lib/resolve-value.js#L71-L76

And here is where we recursively resolve `var()` that resolve to more `var()` https://github.com/MadLittleMods/postcss-css-variables/blob/fe5c821b8feedbe80d8f4399a2b80ca166bb3515/lib/resolve-value.js#L92

It's still unclear to me why that can't be adapted to cover this new case. We may need some sort of `balanced-match` sort of dependency for proper parenthesis pairs but the overall code structure seems like it doesn't need to change to a `while` (this is just masking the bug in the current code)

juliovedovatto replied:
Yeah, I understand your point about the while. It seems redundant in some cases.

---

I tried `balanced-match` before, but the lib does not have a recursive feature. The reason why I choose `match-recursive` lib is that the lib returns an array of what was found.

First, let's assume this particular case:

```css
:root {
  --some-width: 150px;
  --some-other-width: 10px
}
.box-foo {
  width: calc(var(--some-width) - var(--some-other-width)); 
}
```

Using `balanced-match`, the returned output is:
```css
.box-foo {
	width: calc(150px - var(--some-other-width));
}
```

Using `match-recursive`, the returned output is:
```css
.box-foo {
	width: calc(150px - 10px);
}
```

To make these test I used the following functions to match:

```js
var balanced = require('balaced-match');
var matchRecursive = require('match-recursive');

balanced(/\bvar\(/i, /\)/, resultantValue);

matchRecursive(resultantValue, 'var(...)'))

```


It seems `balanced-match` returns only the **first occurrence** of a `var()` declaration. Meanwhile, if I use `match-recursive` lib, I can replace the whole sentence and retrieve the expected value.  Which leads me to understand that `balanced-match` would be the best choice for simple cases of nested functions, whereas `match-recursive` is best for complex cases of `var()` use, with more than one `var()` in the same sentence.

MadLittleMods replied:
I am starting to understand the new `while` loop now. It's about matching `var()`, not resolving. Sorry it took me a while to get my head around it properly.

I'd be more comfortable using [`balanced-match`](https://www.npmjs.com/package/balanced-match) given it's ubiquity. We can also use it like we were before going one by one AFAICT

```js
var balanced = require('balanced-match');

// not tested
while (match = balanced('var(', ')', resultantValue)) {
```

Seems like we should also use this same looping strategy to populate `variablesUsedInValueMap` just above

juliovedovatto replied:
I made changes and explained a little in https://github.com/MadLittleMods/postcss-css-variables/pull/97#issuecomment-487770935

Could you check?

MadLittleMods commented on test/fixtures/nested-inside-other-func.css:
```diff
@@ -1,3 +1,15 @@
+
+.box-foo {
+	width: calc(58.3333333333% - var(--missing, 100px)); 
+}
```
Let's add a test without a fallback

```css
.box-foo {
	width: calc(58.3333333333% - var(--missing)); 
}
```

Also one with multiple `var()`

```css
.box-foo {
	width: calc(var(--some-width) - var(--some-other-width)); 
}
```

juliovedovatto replied:
extra tests added ;)

juliovedovatto commented:
@MadLittleMods I made more code changes, just to adapt your logic in some special scenarios.

I was facing issues with `RegExp` to match few nested inside css functions. 

I spent a good time trying to create a good regex to help, but due the fact JS `RegExp` engine is limited, I had to change my approach.

The `match-recursive` npm lib was added, to help match several `var()` declaration in the same line. Additionally, I changed a few chunks of the code to fit my logic into your code.

I added more related tests and worked ok.

Could you take a look and make some tests? I believe this is a good candidate to merge 🤓

Cheers!

juliovedovatto commented:
@MadLittleMods I did code changes again, please review. I think we are in the right track :)

Now, it will use a `while` loop and `balanced-match` npm lib to match `var()` declarations and resolve them. This should cover nested CSS functions.

In addition, I added one more test for css nested functions. It should be all covered now.

Cheers!

MadLittleMods commented on package.json:
```diff
@@ -13,7 +13,9 @@
     "type": "git",
     "url": "https://github.com/MadLittleMods/postcss-css-variables.git"
   },
+  "preinstall": "npm install escape-string-regexp extend match-recursive postcss",
```
Why was this added?

juliovedovatto replied:
We faced a particular problem with missing packages, o I had to use `preinstall` instruction to force add the missing packages. I removed it now.

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -41,12 +39,19 @@ var resolveValue = function(decl, map, /*optional*/ignorePseudoScope, /*internal
-		// By best match, we mean, the variable actually applies. Criteria:
-		//		- is under the same scope
-		//		- The latest defined `!important` if any
-		var matchingVarDeclMapItem;
```
No need to remove the comment or change the style here

juliovedovatto replied:
re-added, I deleted it by mistake

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -41,12 +39,19 @@ var resolveValue = function(decl, map, /*optional*/ignorePseudoScope, /*internal
+	// matches `name[, fallback]`, captures "name" and "fallback"
+	// See: http://dev.w3.org/csswg/css-variables/#funcdef-var
+	var match;
+	while (match = balanced('var(', ')', resultantValue)) {
```
> Seems like we should also use this same looping strategy to populate variablesUsedInValueMap just above
> 
> https://github.com/MadLittleMods/postcss-css-variables/pull/97#discussion_r279523067

We could also move `RE_VAR_FUNC` to `index.js` then

juliovedovatto replied:
I think it would be good to have a modified version of `RE_VAR_FUNC`, just to check if str have `var(--somethig)` declaration. Remove it would generate an extra work, refactoring `index.js`. I'm not sure if best option for a small fix to match nested css functions.

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -41,12 +39,19 @@ var resolveValue = function(decl, map, /*optional*/ignorePseudoScope, /*internal
+		var matchingVarDeclMapItem = undefined;
+
+		// split comma to find variable name and fallback value
+		match.body = match.body.split(',');
```
Let's make this more robust and not assign back to itself

```js
// Split at the comma to find variable name and fallback value
// There may be other commas in the values so this isn't necessary just 2 pieces
var variableFallbackSplitPieces = match.body.split(',');
```

juliovedovatto replied:
done, changed it for better reading

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -41,12 +39,19 @@ var resolveValue = function(decl, map, /*optional*/ignorePseudoScope, /*internal
+		// split comma to find variable name and fallback value
+		match.body = match.body.split(',');
+		// get variable name and fallback, filtering empty items
+		var [ variableName, fallback ] = [ match.body.shift(), match.body.join(',') ].map(item => item.trim()).filter(item => !!item);
```
Let's avoid the destructuring syntax just to keep with the maybe Node.js compatibility some people like to have.

This also has a bit of necessary array magic. We should opt for something easier to understand.

Is this the same? Seems like we don't need to trim things  and can leave peoples formatting alone.

```js
const variableName = variableFallbackSplitPieces[0].trim();
const fallback = variableFallbackSplitPieces.slice(1).join(',');
```

juliovedovatto replied:
done, changed it for better reading

MadLittleMods commented on test/fixtures/nested-inside-other-func.css:
```diff
@@ -1,3 +1,54 @@
+
+.box-foo {
+	background-color: color(var(--missing, white));
+}
```
Let's split these cases out to separate fixture files so the output is easier to understand when it fails

 - `nested-inside-other-func-with-fallback`
 - `nested-inside-calc-func`
 - `nested-inside-calc-func-with-fallback`
 - `nested-inside-calc-func-with-fallback-var`
 - etc

juliovedovatto replied:
Done. Added more use-case tests

juliovedovatto commented:
@MadLittleMods Another code review round done. Please check.

juliovedovatto commented:
@MadLittleMods any updates?

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -27,26 +29,39 @@ function toString(value) {
-	var variablesUsedInValue = Object.keys(variablesUsedInValueMap);
+	// match all variables first, to gather all available variables in decl.value
+	var variablesUsedInValue = [];
+	while ((matchingVarDecl = RE_VAR_FUNC_G.exec(resultantValue))) {
```
Why are we using a different matching strategy from below? Seems like we should use the same strategy to get the `variableName` as below

juliovedovatto replied:
Reason: I ended up maintaining this strategy, to give more purpose to `RE_VAR_FUNC` RegExp.

Check it now, I changed the matching strategy, to use `balanced` npm lib as well.

MadLittleMods replied:
Seems good now. Both use the same looping strategy

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -27,26 +29,39 @@ function toString(value) {
+		var matchingVarDeclMapItem = undefined;
+
+		// Split at the comma to find variable name and fallback value
+		// There may be other commas in the values so this isn't necessary just 2 pieces
```
*necessarily

Spelled it wrong in my comment before, https://github.com/MadLittleMods/postcss-css-variables/pull/97#discussion_r282411060

juliovedovatto replied:
fix'd

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -27,26 +29,39 @@ function toString(value) {
-		variablesUsedInValueMap[variableName] = true;
-	});
-	var variablesUsedInValue = Object.keys(variablesUsedInValueMap);
+	// match all variables first, to gather all available variables in decl.value
```
Let's capitalize to match the rest of the comments. This also doesn't provide any context for why

```
// Match all variables first so we can later on if there are circular dependencies
```

juliovedovatto replied:
fix'd

colinhowells commented:
Thanks everyone for your hard work on this – I've been watching this bit of the [HiQ framework](https://github.com/jonathanharrell/hiq/blob/9ff1977b1f49d441de599eb8a079f942cf23a0e0/css/elements/forms/radio-buttons.css#L10)
```css
--radio-border-width: var(--hiq-radio-border-width, 1px) solid var(--hiq-radio-border-color, transparent);
--radio-border-color: var(--hiq-radio-border-color, transparent);
border: var(--radio-border-width) solid var(--radio-border-color);
```
get compiled to:
```css
border: 1px) solid var(--hiq-radio-border-color, transparent solid transparent;
```
and had no clue why. If it was defeated by regex, that'd make sense ...

juliovedovatto commented on lib/resolve-value.js:
```diff
@@ -27,26 +29,39 @@ function toString(value) {
+	// match all variables first, to gather all available variables in decl.value
+	var variablesUsedInValue = [];
+	while ((matchingVarDecl = RE_VAR_FUNC_G.exec(resultantValue))) {
+		variablesUsedInValue.push(matchingVarDecl[1]);
```
I saw `variablesUsedInValueMap` variable wasn't being used for anything else, then I tried to re-create the same logic, but using a `while` loop. I can rollback this change, if you prefer.

MadLittleMods replied:
I prefer the previous version

juliovedovatto commented:
@MadLittleMods Done with the requested changes. Could you take a look? :)

@colinhowells  You are right, it was defeated by regex. The intend of my proposal is to fix when some complex `var()` usage, specially nested appear.

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -27,26 +29,61 @@ function toString(value) {
+	// Match all variables first so we can later on if there are circular dependencies
+	var variablesUsedInValue = [];
+	// Create a temporary variable, storing resultantValue variable value
+	var resultantValueTemp = resultantValue;
```
`resultantValueTemp` -> `remainingVariableValue` to be a bit more descriptive

Since we keep clearing it out as go

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -27,26 +29,61 @@ function toString(value) {
+	resultantValueTemp = undefined;
+
+	// Old strategy to find used variable names in declaration value, using RegExp
+	// TODO: remove unused block 
```
TODO: remove

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -27,26 +29,61 @@ function toString(value) {
 	var debugIndent = _debugIsInternal ? '\t' : '';
 
+	var matchingVarDecl = undefined;
+	var RE_VAR_FUNC_G = new RegExp(RE_VAR_FUNC.source, 'g');
```
We can get rid of `RE_VAR_FUNC_G` and `RE_VAR_FUNC` now

juliovedovatto replied:
`RE_VAR_FUNC_G` was removed.

But, It seems we can't remove `RE_VAR_FUNC`, since it is being called in `index.js`. 

https://github.com/MadLittleMods/postcss-css-variables/blob/5695b77f5b7b1bca7f570ef1b1fb324fc0f432f7/index.js#L216-L227

juliovedovatto commented:
@MadLittleMods I did some of your requests, could you review again?

Also, check my comment https://github.com/MadLittleMods/postcss-css-variables/pull/97#discussion-diff-296426364R32

juliovedovatto commented:
@MadLittleMods any news?

juliovedovatto commented:
@MadLittleMods sorry to bump this PR, but could you take a look in the latest changes sent?

MadLittleMods commented on lib/resolve-value.js:
```diff
@@ -1,21 +1,23 @@
 
+function filterDistinct(value, index, self) {
+	return self.indexOf(value) === index;
+}
```
We can get rid of this unused function

MadLittleMods commented:
I've left this for a while. Just feeling the effects of going back and fourth with things not actually changing or being quite finished.

I only spotted one thing and the tests are passing so we can merge after that 😥

juliovedovatto commented:
No worries @MadLittleMods 

I just did the request change. Feel free to to approve ;)

MadLittleMods commented:
Thanks for the quick update @juliovedovatto ❤️ 

This is now part of `postcss-css-variables@0.14.0` :tada:, https://github.com/MadLittleMods/postcss-css-variables/blob/master/CHANGELOG.md#v0140---2019-11-24
